File: api/src/internal/scoring/attributes.ts
import * as normalize from "./normalize";

interface AttributeMix {
  Activity?: number;
  Projects?: number;
  OpenSource?: number;
  Popularity?: number;
  Consistency?: number;
  Technical?: number;
}

export interface Attributes {
  pac: number; // Development Speed
  sho: number; // Project Execution
  pas: number; // Collaboration
  dri: number; // Technical Versatility
  def: number; // Code Discipline
  phy: number; // Developer Consistency
}

function applyMix(c: { Activity: number; Projects: number; OpenSource: number; Popularity: number; Consistency: number; Technical: number }, mix: AttributeMix): number {
  const val =
    c.Activity * (mix.Activity || 0) +
    c.Projects * (mix.Projects || 0) +
    c.OpenSource * (mix.OpenSource || 0) +
    c.Popularity * (mix.Popularity || 0) +
    c.Consistency * (mix.Consistency || 0) +
    c.Technical * (mix.Technical || 0);
  return Math.floor(normalize.clamp(val, 0, 100) + 0.5);
}

export function computeAttributes(c: { Activity: number; Projects: number; OpenSource: number; Popularity: number; Consistency: number; Technical: number }): Attributes {
  return {
    pac: applyMix(c, { Activity: 0.70, Consistency: 0.20, Projects: 0.10 }),
    sho: applyMix(c, { Projects: 0.65, Activity: 0.20, Popularity: 0.15 }),
    pas: applyMix(c, { OpenSource: 0.85, Popularity: 0.15 }),
    dri: applyMix(c, { Technical: 0.80, Projects: 0.20 }),
    def: applyMix(c, { Projects: 0.45, Consistency: 0.35, OpenSource: 0.20 }),
    phy: applyMix(c, { Consistency: 0.70, Activity: 0.30 }),
  };
}

File: api/internal/scoring/attributes.go
package scoring

// Attributes holds the six FIFA/EA-FC-style stats, each 0-100.
type Attributes struct {
	PAC int `json:"pac"` // Development Speed
	SHO int `json:"sho"` // Project Execution
	PAS int `json:"pas"` // Collaboration
	DRI int `json:"dri"` // Technical Versatility
	DEF int `json:"def"` // Code Discipline
	PHY int `json:"phy"` // Developer Consistency
}

func applyMix(c CategoryScores, mix attributeMix) int {
	val := c.Activity*mix.Activity +
		c.Projects*mix.Projects +
		c.OpenSource*mix.OpenSource +
		c.Popularity*mix.Popularity +
		c.Consistency*mix.Consistency +
		c.Technical*mix.Technical
	return int(clamp(val, 0, 100) + 0.5) // round to nearest int
}

// ComputeAttributes maps category scores onto the six named attributes using the
// mixes defined in weights.go.
func ComputeAttributes(c CategoryScores) Attributes {
	return Attributes{
		PAC: applyMix(c, pacMix),
		SHO: applyMix(c, shoMix),
		PAS: applyMix(c, pasMix),
		DRI: applyMix(c, driMix),
		DEF: applyMix(c, defMix),
		PHY: applyMix(c, phyMix),
	}
}
